CONTRIBUTING.md

```diff
@@ -49,15 +49,106 @@ If your pull request contains any of the below, it's less likely to be merged:
+- [`IUpgradeContextFactory`](src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/IUpgradeContextFactory.cs) is responsible for creating new instances of [`IUpgradeContext`](src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/IUpgradeContext.cs) for the Upgrade Assistant to use. The upgrade context (described below) is the primary object storing the current state of the upgrade proces. It includes information about the current entry point and current projects being upgraded and APIs for accessing `IProject` and `IProjectFile` abstractions for the projects being upgraded which allow querying or modifying the projects' properties.
+- [`IUpgradeStateManager`](src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/IUpgradeStateManager.cs) (implemented, by default, by [`FileUpgradeStateFactory`](src/cli/Microsoft.DotNet.UpgradeAssistant.Cli/FileUpgradeStateFactory.cs)) is responsible for storing and reading state that is preserved between runs of Upgrade Assistant in a given directory. The state file that it reads from and writes to disk (.upgrade-assistant) contains the current project being upgraded, the current entry point, and a property bag that can hold arbitrary state (but is currently only used to store the base backup location a user prefers). This state can be useful for Upgrade Assistant to start up more quickly but should not be *necessary*.
+- Upgrade startup types (implementing [`IUpgradeStartup`](src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/IUpgradeStartup.cs)) are components that need to be executed once as Upgrade Assistant starts up to run pre-commands or otherwise configure settings that will be needed by other components as the tool executes. These startup tasks are run by the `ConsoleRunner` prior to creating the upgrade context or initializing or running any upgrade steps. Current built-in startup implementations include:
+  - [`MSBuildRegistrationStartup`](src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/MSBuildRegistrationStartup.cs) which registers the MSBuild instance on the user's computer that will be used while running Upgrade Assistant. Upgrade Assistant does not ship with any MSBUild or NuGet binaries. Instead, the versions of these binaries present on the user's machine are used so that the build processes executed during Upgrade Assistant resemble the user's typical building of their project as much as possible.
```

Comment:
I wouldn't say these are "core" components

Reply:
The upgrader manager is pretty core. The context factory, state manager, and upgrade startup types probably less so. They seem like useful concepts to understand, but maybe I can put them in a 'other concepts' section that just gives one-line descriptions and links to source.